barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_alu_trace.cpp

```diff
@@ -731,4 +743,218 @@ FF AvmAluTraceBuilder::op_cast(FF const& a, AvmMemoryTag in_tag, uint32_t clk)
+            .shift_lt_bit_len = false,
+        });
+        return 0;
+    }
+    // We decompose the input into two limbs partitioned at the b-th bit, we use x_lo and x_hi
+    // to avoid any confusion with the a_lo and a_hi that form part of the range check
+    auto [x_lo, x_hi] = decompose(a, b_u8);
+    // We can skip the dynamic range check by performing a static one
+    // rng_chk_lo = 2^b - x_lo && rng_chk_hi = 2^(num_bits - b) - x_hi
+    uint256_t rng_chk_lo = (uint256_t(1) << b_u8) - x_lo;
+    uint256_t rng_chk_hi = (uint256_t(1) << (num_bits - b_u8)) - x_hi;
+
+    // Each hi and lo limb is range checked over 128bits
+    uint256_t limb = rng_chk_lo + (rng_chk_hi << uint256_t(128));
+    // Load the range check values into the ALU registers
+    auto [alu_u8_r0, alu_u8_r1, alu_u16_reg] = AvmAluTraceBuilder::to_alu_slice_registers(limb);
+
+    // Add counters for the pow of two lookups
+    u8_pow_2_counters[1][num_bits - b_u8]++;
+
+    alu_trace.push_back(AvmAluTraceBuilder::AluTraceEntry{
+        .alu_clk = clk,
+        .alu_op_shr = true,
+        .alu_u8_tag = in_tag == AvmMemoryTag::U8,
+        .alu_u16_tag = in_tag == AvmMemoryTag::U16,
+        .alu_u32_tag = in_tag == AvmMemoryTag::U32,
+        .alu_u64_tag = in_tag == AvmMemoryTag::U64,
+        .alu_u128_tag = in_tag == AvmMemoryTag::U128,
+        .alu_ia = a,
+        .alu_ib = b,
+        // Could be replaced with x_hi but nice to have 2 ways of calculating the result
+        .alu_ic = FF(c_u256),
+        .alu_u8_r0 = alu_u8_r0,
+        .alu_u8_r1 = alu_u8_r1,
+        .alu_u16_reg = alu_u16_reg,
+        .hi_lo_limbs{ rng_chk_lo, rng_chk_hi, x_lo, x_hi },
+        .mem_tag_bits = num_bits,
+        .mem_tag_sub_shift = static_cast<uint8_t>(num_bits - b_u8),
+        .shift_lt_bit_len = true,
+
+    });
+    return c_u256;
+}
+
+/**
+ * @brief Build Alu trace and compute the result of a SHL operation on two operands of type defined by in_tag.
+ *
+ * @param a Left operand of the SHL
+ * @param b Right operand of the SHL
+ * @param clk Clock referring to the operation in the main trace.
+ * @param in_tag Instruction tag defining the number of bits for the SHL.
+ *
+ * @return FF The boolean result of SHL casted to a finite field element
+ */
+FF AvmAluTraceBuilder::op_shl(FF const& a, FF const& b, AvmMemoryTag in_tag, uint32_t clk)
+{
+    // Perform the shift operation over 256-bit integers
+    uint256_t a_u256{ a };
+    // Check that the shift amount is an 8-bit integer
+    ASSERT(uint256_t(b) < 256);
+    uint8_t b_u8 = static_cast<uint8_t>(uint256_t(b));
+
+    uint256_t c_u256 = a_u256 << b_u8;
+
+    uint8_t num_bits = mem_tag_bits(in_tag);
+    u8_pow_2_counters[0][b_u8]++;
+    // If we are shifting more than the number of bits, the result is trivially 0
+    if (b_u8 >= num_bits) {
+        u8_pow_2_counters[1][b_u8 - num_bits]++;
+        [[maybe_unused]] auto unused_regs = AvmAluTraceBuilder::to_alu_slice_registers(0);
```

Comment:
i update this with a comment, we can't because this function also updates the counter for the lookups involving the range check of zero.

There are two things we could do  in the future.
(1) Decouple the counter incrementing logic for range checks from the function `to_alu_slice_registers`
(2) Possibly optimise the pil relation to ignore a range check in this situation (i.e. shift_lt_bin_len = 0)

Reply:
Ok. Then a comment is definitely welcome.